src/Symfony/Component/HttpClient/ResponseSetMonitor.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Monitors a set of responses and triggers callbacks as they complete.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ResponseSetMonitor
+{
+    private $client;
+    private $responses;
+
+    public function __construct(HttpClientInterface $client)
+    {
+        $this->client = $client;
+        $this->responses = new \SplObjectStorage();
+    }
+
+    public function add(ResponseInterface $response, callable $onHeaders = null, callable $onBody = null, callable $onError = null): void
+    {
+        $this->responses[$response] = [$onHeaders, $onBody, $onError];
+        $this->tick();
+    }
+
+    public function tick(?float $idleTimeout = 0.0): int
```

Comment:
I read the code in master, and passing `null` means `INF`. IMHO there is no need to all passing explicitly `null` here.
I would remove `?` in the typehint.

Reply:
`null` actually means to use the default request's timeout.
But I still removed the `?` and I changed the meaning of the timeout: now it's a "max wait time", not only a "max idle time".